[Client-gen] Make client generator based on go2idl

caesarxuchao opened the pull request:
The generator in this PR will generate client code for top-level types, e.g., the code in pkg/client/unversioned/deployments.go.

ref #17468

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/17534)

<!-- Reviewable:end -->

Commit:
First step of adding clinet-gen: add the generator that generates client code
for top-level types.

caesarxuchao commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,283 @@
+		glog.Fatalf("Failed loading boilerplate: %v", err)
+	}
+
+	// @lavalamp: This is quite hacky. Any suggestions?
```
I "@lavalamp" in the code where I'm not sure what the best practice is.

wojtek-t replied:
Just to ensure that we are on the same page. You would like to generate clients for all "internal" types (i.e. api.Pod, api.ReplicationController, ..., extensions.Deployment, extensions.Job, ...).

So what I would do is:
- in the main binary overwrite the InputDirs:
https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/set-gen/main.go#L39
- it should container list of all packages, where your types are defined, i.e.:
k8s.io/kubernetes/pkg/api
k8s.io/kubernetes/pkg/apis/extensions
(possibly some others)
- for every type you would like to generate the types, add the following comment line next to it:
// +genclient=true
- then here, iterate over all types in the context and filter only those that have the genclient "bit" set

This exact pattern is used in set-gen example, in particular the filtering is done here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/set-gen/generators/sets.go#L115

caesarxuchao replied:
> Just to ensure that we are on the same page. You would like to generate clients for all "internal" types (i.e. api.Pod, api.ReplicationController, ..., extensions.Deployment, extensions.Job, ...)

Yes, that's what I'm trying to do.

Thanks for the suggestion. That sounds much better. Let me give it a try.

lavalamp commented:
@krousey can you take a first pass on this while I build-cop?

krousey commented on cmd/libs/go2idl/make-client/generators/imported_types.go:
```diff
@@ -0,0 +1,56 @@
+	// We use these types in the template to trigger the ImportTracker
+	fieldSelector = &types.Type{
+		Name: types.Name{
+			"k8s.io/kubernetes/pkg/fields",
```
Please use `Name:` and `Package:` when using struct literals from other packages.

caesarxuchao replied:
Ok.

krousey commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,283 @@
+		HeaderText: append(boilerplate, []byte(
+			`
+// This file was autogenerated by the command:
+// $ `+strings.Join(os.Args, " ")+`
```
I'm not certain, but shouldn't there be a right delimiter here?

krousey replied:
Oh I see. This isn't a template.

krousey commented on cmd/libs/go2idl/make-client/output/doc.go:
```diff
@@ -0,0 +1,22 @@
+/*
```
This is silly to put in a separate file. Can we just put it in the one output file?

caesarxuchao replied:
I think we have a rule requiring a doc.go for each package, though not sure why we have the rule.

krousey replied:
@lavalamp Do we have that rule? It seems weird to have a doc.go file when there is only one or mainly one file for the package.

caesarxuchao replied:
According to the client proposal, in the future there will be files for each resource in that group/version. Will that mitigate the uglyness?

lavalamp replied:
maybe? I think it's better to have it? If nothing else it's where to put
package-godoc, which we do definitely want.

On Fri, Nov 20, 2015 at 4:03 PM, krousey <notifications@github.com> wrote:

> In cmd/libs/go2idl/make-client/output/doc.go
> <https://github.com/kubernetes/kubernetes/pull/17534#discussion_r45534464>
> :
>
> > @@ -0,0 +1,22 @@
> > +/*
>
> @lavalamp <https://github.com/lavalamp> Do we have that rule? It seems
> weird to have a doc.go file when there is only one or mainly one file for
> the package.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17534/files#r45534464>.
>

krousey commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,283 @@
+// We could use t directly as the argument, but doing it this way makes it easy
+// to mix in additional parameters. This feature is not used in this set
+// generator, but is present as an example.
+func (g *genClient) args(t *types.Type, kv ...interface{}) interface{} {
```
Is this needed? Why can't you just define this map inline in GenerateType and pass it to all the templates?

caesarxuchao replied:
Good point. I think we can. I just copied this function from Daniel's set-gen example without putting much thought into it.

krousey commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,283 @@
+
+// GenerateType makes the body of a file implementing a set for type t.
+func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
```
The standard template delimiters are {{ and }} respectively. Is there a really compelling reason to stray from the standard template syntax that most Go developers are used to?

caesarxuchao replied:
I guess it's to allow people to write the standard template in this template without being interpreted by go2idl?

@lavalamp since he uses "$" in his example.

lavalamp replied:
Yes. {{ doesn't parse well when next to a go-language {.

On Fri, Nov 20, 2015 at 2:21 PM, krousey <notifications@github.com> wrote:

> In cmd/libs/go2idl/make-client/generators/client-generator.go
> <https://github.com/kubernetes/kubernetes/pull/17534#discussion_r45526533>
> :
>
> > +	// @lavalamp: how about making a plural function in pkg/util, RESTMapper also needs such a function.
> > +	// m["plural"] = util.Plural(type.Name.Name)
> > +	m["fieldSelector"] = fieldSelector
> > +	m["labelSelector"] = labelSelector
> > +	m["watchInterface"] = watchInterface
> > +	m["apiDeleteOptions"] = apiDeleteOptions
> > +	m["apiListOptions"] = apiListOptions
> > +	for i := 0; i < len(kv)/2; i++ {
> > +		m[kv[i*2]] = kv[i*2+1]
> > +	}
> > +	return m
> > +}
> > +
> > +// GenerateType makes the body of a file implementing a set for type t.
> > +func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
> > +	sw := generator.NewSnippetWriter(w, c, "$", "$")
>
> The standard template delimiters are {{ and }} respectively. Is there a
> really compelling reason to stray from the standard template syntax that
> most Go developers are used to?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17534/files#r45526533>.
>

wojtek-t replied:
As @lavalamp - we already had this discussion and decided that we will be using "$" in our generators.

krousey commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,283 @@
+// GenerateType makes the body of a file implementing a set for type t.
+func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	sw.Do(nameSpacerTemplate, g.args(t))
```
Why are we waiting until this point to try to parse the templates? I would think it'd be better to use template.MustParse at the package level definition so you wouldn't have to have a unit test hit these lines to know you have template errors.

caesarxuchao replied:
We cannot tell if the template is legal without knowing the 1. the definition of the argument (`g.args(t)`), 2. if the namer exists (e.g., `public`).

I think we really need unit test to tell if the template is generating what we expected.

krousey replied:
Templates have a syntax that must be parsable without having to go through
processing. See the standard library's template.MustParse and template.Parse
On Nov 20, 2015 14:44, "Chao Xu" <notifications@github.com> wrote:

> In cmd/libs/go2idl/make-client/generators/client-generator.go
> <https://github.com/kubernetes/kubernetes/pull/17534#discussion_r45528626>
> :
>
> > +	// m["plural"] = util.Plural(type.Name.Name)
> > +	m["fieldSelector"] = fieldSelector
> > +	m["labelSelector"] = labelSelector
> > +	m["watchInterface"] = watchInterface
> > +	m["apiDeleteOptions"] = apiDeleteOptions
> > +	m["apiListOptions"] = apiListOptions
> > +	for i := 0; i < len(kv)/2; i++ {
> > +		m[kv[i*2]] = kv[i*2+1]
> > +	}
> > +	return m
> > +}
> > +
> > +// GenerateType makes the body of a file implementing a set for type t.
> > +func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
> > +	sw := generator.NewSnippetWriter(w, c, "$", "$")
> > +	sw.Do(nameSpacerTemplate, g.args(t))
>
> We cannot tell if the template is legal without knowing the 1. the
> definition of the argument (g.args(t)), 2. if the namer exists (e.g.,
> public).
>
> I think we really need unit test to tell if the template is generating
> what we expected.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17534/files#r45528626>.
>

lavalamp replied:
Functions are injected in the .Do method. There's no benefit to parsing
early since they won't break once you get them working, and they're not
reused, and go2idl is not slow anyway.

On Fri, Nov 20, 2015 at 2:23 PM, krousey <notifications@github.com> wrote:

> In cmd/libs/go2idl/make-client/generators/client-generator.go
> <https://github.com/kubernetes/kubernetes/pull/17534#discussion_r45526693>
> :
>
> > +	// m["plural"] = util.Plural(type.Name.Name)
> > +	m["fieldSelector"] = fieldSelector
> > +	m["labelSelector"] = labelSelector
> > +	m["watchInterface"] = watchInterface
> > +	m["apiDeleteOptions"] = apiDeleteOptions
> > +	m["apiListOptions"] = apiListOptions
> > +	for i := 0; i < len(kv)/2; i++ {
> > +		m[kv[i*2]] = kv[i*2+1]
> > +	}
> > +	return m
> > +}
> > +
> > +// GenerateType makes the body of a file implementing a set for type t.
> > +func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
> > +	sw := generator.NewSnippetWriter(w, c, "$", "$")
> > +	sw.Do(nameSpacerTemplate, g.args(t))
>
> Why are we waiting until this point to try to parse the templates? I would
> think it'd be better to use template.MustParse at the package level
> definition so you wouldn't have to have a unit test hit these lines to know
> you have template errors.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17534/files#r45526693>.
>

krousey commented on cmd/libs/go2idl/make-client/output/deployment.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// deployments implements DeploymentInterface
+type deployments struct {
```
I envisioned these types being public and `newDeployments` public as well.

caesarxuchao replied:
The original file is here: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/deployment.go#L50

They are private in the original file. I'll check why they are private.

krousey commented on cmd/libs/go2idl/make-client/output/deployment.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// newDeployments returns a Deployments
+func newDeployments(c *ExtensionsClient, namespace string) *deployments {
```
Shouldn't this take a RESTClient?

caesarxuchao replied:
Oh, I haven't taken the client proposal into consideration yet. I'm just trying to replicate the original files in pkg/client/unversioned/ at this point.

wojtek-t replied:
I agree with @krousey that it should be RESTClient.
But I support @caesarxuchao here - we should start with replicating what we have, and then it will be so much easier to change it.
So I would leave it as is for now (possibly adding a TODO to fix it).

caesarxuchao replied:
Taking a closer look of how `newDeployments` is used, I'm not sure why it's necessary to change to RESTClient. @krousey could you develop the point? Thanks.

krousey commented on cmd/libs/go2idl/make-client/output/deployment.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// DeploymentInterface has methods to work with Deployment resources.
+type DeploymentInterface interface {
```
What purpose do these interfaces serve? Shouldn't clients just use `deployments` directly?

kargakis commented on cmd/libs/go2idl/make-client/output/deployment.go:
```diff
@@ -0,0 +1,118 @@
+
+// DeploymentNamespacer has methods to work with Deployment resources in a namespace
+type DeploymentNamespacer interface {
+	Deployment(namespace string) DeploymentInterface
```
[s/Deployment/Deployments](https://github.com/kubernetes/kubernetes/blob/3fbf0cb8100a8cf03cf3a077ebc5987fe1b6a5c8/pkg/client/unversioned/deployment.go#L29)

caesarxuchao replied:
Good catch. Thanks!

caesarxuchao commented:
@krousey, I addressed some of your comments. I plan to make the generated the client the same as our current client, and change it according to the client package proposal later, so some of your comments are not addressed. 
Do you think this PR can be merged on its own? I will make following PR to 1. generate clients for other resources, 2. add a test type.

caesarxuchao commented:
@wojtek-t could you take a look as well? Thanks.

wojtek-t commented:
@caesarxuchao - I can take a look tomorrow

krousey commented:
@caesarxuchao As a first step, yea this could be checked in. There are some other comments open that don't seem to be addressed. Also the code still has some @lavalamp in it. That should be removed.

caesarxuchao commented:
@wojtek-t, I marked `@lavalamp` in the code where I'm not sure what's the best go2idl practice, maybe you can take a look tomorrow. Thanks.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+limitations under the License.
+*/
+
+// Package generators has the generators for the set-gen utility.
```
s/set-gen/client-gen/

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+/*
```
For consistency with set-gen and deepcopy-gen, I would change the directory name from:
cmd/libs/go2idl/make-client/
to
cmd/libs/go2idl/client-gen/

caesarxuchao replied:
Sure.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+	for _, m := range ref.Members {
+		topLevelTypes[m.Type] = struct{}{}
+	}
+	return generator.Packages{&generator.DefaultPackage{
```
Regarding my comment above - what you would like to do, is exactly the same what in set-gen example is done.
So you
- generate "doc.go" file:
https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/set-gen/generators/sets.go#L71
- for every type (that is not filtered out), you generate a separate file, creating a dedicated generator for it:
https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/set-gen/generators/sets.go#L83

wojtek-t replied:
Sorry - it seems you're already doing this :)

wojtek-t replied:
So I would just change the filter function.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+func (g *genClient) Imports(c *generator.Context) (imports []string) {
+	return append(g.imports.ImportLines())
```
nit: if you don't need to add any "additional" imports, just return g.imports.ImportLines()

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	pkg := filepath.Base(t.Name.Package)
+	m := map[interface{}]interface{}{
```
why not map[string]interface{} ?

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+		"type":    t,
+		"package": pkg,
+		"Package": namer.IC(pkg),
+		// @lavalamp: how about making a plural function in pkg/util, RESTMapper also needs such a function.
```
I'm not sure I understand. You want to have a function, that e.g. transforms "Job" to "Jobs"?

I'm fine with that, but I would put it in "namer" package.

caesarxuchao replied:
I meant to ask if it's worth to add such a namer. 

`RESTMapper` is a separate problem, I should have removed it from this comment.

wojtek-t replied:
Yes - I think it's worth adding.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+		"Package": namer.IC(pkg),
+		// @lavalamp: how about making a plural function in pkg/util, RESTMapper also needs such a function.
+		// m["plural"] = util.Plural(type.Name.Name)
+		"fieldSelector":    fieldSelector,
```
I would like to avoid explicitly declaring those types. and I think it is doable.
What you probably want to do is to:
- extend InputDir param with packages where those types are defined (e.g. pkg/fields)
- here use the given context:
    "c.Universe.Get(types.Name("k8s.io/kubernetes/pkg/fields", "Selector")"

[I didn't check it, but I'm pretty sure this should work.]

caesarxuchao replied:
Sure. That should work.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/imported_types.go:
```diff
@@ -0,0 +1,57 @@
+/*
```
I would like to avoid having this file - see my comment above where you are using these vars.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+		"apiDeleteOptions": apiDeleteOptions,
+		"apiListOptions":   apiListOptions,
+	}
+	sw.Do(nameSpacerTemplate, m)
```
why not "namespacerTemplate"?

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+	sw.Do(interfaceTemplate, m)
+	sw.Do(structTemplate, m)
+	sw.Do(newStructTemplate, m)
+	sw.Do(listTemplate, m)
```
nit: to be honest, I would change the order of the methods to:
Create
Update
Delete
Get
List
Watch
// possibly others?

I know why you have that order (it's the same as in Deployment), but we have a different order in every file, so we need to unify it anyway, and if we change that anyway, having something more intuitive would be nice.
WDYT?

caesarxuchao replied:
Sure. The order you listed looks good. Thanks.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+var interfaceTemplate = `
+// $.type|public$Interface has methods to work with $.type|public$ resources.
+type $.type|public$Interface interface {
+	List(label $.labelSelector|raw$, field $.fieldSelector|raw$) (*$.type|raw$List, error)
```
Same here (if you decide to change the order of methods above).

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+	sw.Do(deleteTemplate, m)
+	sw.Do(createTemplate, m)
+	sw.Do(updateTemplate, m)
+	sw.Do(updateStatusTemplate, m)
```
I'm not sure what we should do with UpdateStatus - this is still present in only few resources, so that is still not generic.

caesarxuchao replied:
Removed.

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+// List takes label and field selectors, and returns the list of $.type|public$s that match those selectors.
+func (c *$.type|private$s) List(label $.labelSelector|raw$, field $.fieldSelector|raw$) (result *$.type|raw$List, err error) {
+	result = &$.type|raw$List{}
+	err = c.client.Get().Namespace(c.ns).Resource("$.type|private$s").LabelsSelectorParam(label).FieldsSelectorParam(field).Do().Into(result)
```
nit: what do you think about separating this line into few lines (similarly to Watch() function) to make it more readable?

wojtek-t commented on cmd/libs/go2idl/make-client/generators/client-generator.go:
```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return err
+	}
+	return c.client.Delete().Namespace(c.ns).Resource("$.type|private$s").Name(name).Body(body).Do().Error()
```
Same here and in other places.

wojtek-t commented on cmd/libs/go2idl/make-client/input/types.go:
```diff
@@ -0,0 +1,29 @@
+/*
```
If you decide to proceed with my suggestion of using:
// +genclient=true
comment, please remove this file.

wojtek-t commented on cmd/libs/go2idl/make-client/main.go:
```diff
@@ -0,0 +1,42 @@
+limitations under the License.
+*/
+
+// make-client makes the individual typed clients using go2idl.
```
gen-client

wojtek-t commented:
@caesarxuchao - I've added a bunch of comments, but generally it looks good!

One high level comment is that I think that:
- we should start with the PR just creating a generator
- start using this generator to produce a real code, in a separate PR
WDYT?

caesarxuchao commented:
> One high level comment is that I think that:
- we should start with the PR just creating a generator
- start using this generator to produce a real code, in a separate PR

@wojtek-t, If I understand correctly, you suggest that we separate all those templates into a separate PR? That makes sense. Do you want me to do it this time though? As you've reviewed the entire PR anyway :)

wojtek-t commented:
>
@wojtek-t, If I understand correctly, you suggest that we separate all those templates into a separate PR? That makes sense. Do you want me to do it this time though? As you've reviewed the entire PR anyway :)

No - I meant moving the code that is auto-generated to a separate PR.
So in this case I meant moving:
cmd/libs/go2idl/make-client/output/deployment.go
to a separate file (and possibly generate it for everything).

Does it make sense?

caesarxuchao commented:
Sure. Usually I just separate them in commits, sorry I failed to do that this time.

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/generator-for-type.go:
```diff
@@ -0,0 +1,205 @@
+/*
```
I make this a separate file because we will have other generator, like the one for the group client in #17699

wojtek-t replied:
Sure - SGTM

caesarxuchao commented on cmd/libs/go2idl/client-gen/main.go:
```diff
@@ -0,0 +1,48 @@
+	// Override defaults. These are Kubernetes specific input and output
+	// locations.
+	arguments.InputDirs = []string{
+		"k8s.io/kubernetes/pkg/apis/extensions",
```
Change InputDirs according to @wojtek-t's comment

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/generator-for-type.go:
```diff
@@ -0,0 +1,205 @@
+	sw.Do(interfaceTemplate, m)
+	sw.Do(structTemplate, m)
+	sw.Do(newStructTemplate, m)
+	sw.Do(createTemplate, m)
```
Order changed according to @wojtek-t's comment.

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/client-generator.go:
```diff
@@ -0,0 +1,97 @@
+			return generators
+		},
+		FilterFunc: func(c *generator.Context, t *types.Type) bool {
+			return types.ExtractCommentTags("+", t.CommentLines)["genclient"] == "true"
```
FilterFunc changed.

caesarxuchao commented:
@wojtek-t I should have addressed all you comments. PTAL. Thanks.

The output is #17738.

I will make another PR to add plural namer.

wojtek-t commented on cmd/libs/go2idl/client-gen/main.go:
```diff
@@ -0,0 +1,48 @@
+		"k8s.io/kubernetes/pkg/fields",
+		"k8s.io/kubernetes/pkg/labels",
+		"k8s.io/kubernetes/pkg/watch",
+		"k8s.io/kubernetes/pkg/api",
```
Can you please sort these alphabetically?

wojtek-t commented on cmd/libs/go2idl/client-gen/main.go:
```diff
@@ -0,0 +1,48 @@
+		"k8s.io/kubernetes/pkg/watch",
+		"k8s.io/kubernetes/pkg/api",
+	}
+	arguments.OutputPackagePath = "k8s.io/kubernetes/cmd/libs/go2idl/client-gen/output"
```
I think you should remove this one - those should be generated into pkg/client/unversioned, right?

caesarxuchao replied:
I'll change it to pkg/client/clientset/unversioned, and leave a TODO here.

wojtek-t commented on pkg/apis/extensions/types.go:
```diff
@@ -185,6 +185,7 @@ type ThirdPartyResourceData struct {
 	Data []byte `json:"name,omitempty"`
 }
 
+// +genclient=true
```
Can you please move this one to the next PR (where you really generate the code)?

wojtek-t commented on cmd/libs/go2idl/client-gen/generators/client-generator.go:
```diff
@@ -0,0 +1,97 @@
+		glog.Fatalf("Failed loading boilerplate: %v", err)
+	}
+	return generator.Packages{&generator.DefaultPackage{
+		PackageName: "unversionedClient",
```
I think this should be just "unversioned" (assuming that we want to leave those files where they currently are (i.e. pkg/client/unversioned). WDYT?

caesarxuchao replied:
I changed it to be just "unversioned", thanks.

wojtek-t commented:
@caesarxuchao - I've added few, but very minor comments. 
Feel free to self-apply "lgtm" label.

wojtek-t commented:
@caesarxuchao - also - please squash before applying lgtm label.

caesarxuchao commented:
I've addressed the comments and squashed. I'm going to apply the lgtm. Thank you @krousey and @wojtek-t for the valuable suggestions.

wojtek-t commented:
LGTM - thanks!

wojtek-t commented:
@caesarxuchao - one more thing - can you please change the title to more reflect the contents of this PR (especially remove [WIP]).

caesarxuchao commented:
Thank you @wojtek-t. I updated the first comment as well.
